File: apps/codemap-search/experiments/event-navigation-poc/rust_macros.py
"""Expand same-scope, single-ident declarative item macros from their source.

Literal impl templates and restricted identifier-shadowing statement templates
are accepted. Repetitions, general hygiene and procedural execution stay bounds.
"""

from bisect import bisect_right
from copy import copy
import re


def expand_ident_impl_macros(source, parser, max_file_bytes):
    from syntax import child, walk
    definitions = {}
    root = source.tree.root_node
    for node in root.named_children:
        if node.type == "macro_definition":
            definitions.setdefault(source.text(child(node, "name")), []).append(node)
    edits = []
    for statement in root.named_children:
        invocation = statement
        if statement.type == "expression_statement" and len(statement.named_children) == 1:
            invocation = statement.named_children[0]
        if invocation.type != "macro_invocation":
            continue
        macro = source.text(child(invocation, "macro"))
        candidates = [node for node in definitions.get(macro, ()) if node.start_byte < invocation.start_byte]
        if len(candidates) != 1:
            continue
        definition = candidates[0]
        rules = [node for node in definition.named_children if node.type == "macro_rule"]
        if len(rules) != 1:
            continue
        pattern, body = child(rules[0], "left"), child(rules[0], "right")
        match = re.fullmatch(r"\(\s*\$(\w+)\s*:\s*ident\s*\)", source.text(pattern))
        arguments = next((node for node in invocation.named_children if node.type == "token_tree"), None)
        values = [node for node in arguments.named_children if node.type not in {"line_comment", "block_comment"}] if arguments else []
        if not match or len(values) != 1 or values[0].type not in {"identifier", "type_identifier"}:
            continue
        variable, argument = "$" + match[1], values[0]
        replacements = []
        supported = True
        for node in walk(body):
            if "repetition" in node.type:
                supported = False
            if node.type == "metavariable":
                if source.text(node) != variable:
                    supported = False
                replacements.append(node)
        if not supported or not replacements:
            continue
        start, end = body.start_byte + 1, body.end_byte - 1
        chunks = []
        cursor = start
        for variable_node in replacements:
            chunks.append((source.data[cursor:variable_node.start_byte], cursor))
            chunks.append((source.data[argument.start_byte:argument.end_byte], argument.start_byte))
            cursor = variable_node.end_byte
        chunks.append((source.data[cursor:end], cursor))
        expanded = b"".join(data for data, _ in chunks)
        tree = parser.parse(expanded)
        allowed = {"impl_item", "line_comment", "block_comment", "attribute_item"}
        if tree.root_node.has_error or not tree.root_node.named_children or any(node.type not in allowed for node in tree.root_node.named_children):
            continue
        edits.append((statement.start_byte, statement.end_byte, invocation.start_byte, chunks))
    if not edits:
        return 0
    expected_bytes = len(source.data) + sum(sum(len(data) for data, _ in chunks) - (end - start) for start, end, _, chunks in edits)
    if len(edits) > 256 or expected_bytes > max_file_bytes:
        return 0
    chunks, segments, expansions = [], [], []
    length, cursor = 0, 0
    def append(data, origin):
        nonlocal length
        if data:
            segments.append((length, length + len(data), origin))
            chunks.append(data)
            length += len(data)
    for start, end, invocation, expanded_chunks in edits:
        append(source.data[cursor:start], cursor)
        expansion_start = length
        for data, origin in expanded_chunks:
            append(data, origin)
        expansions.append((expansion_start, length, invocation))
        cursor = end
    append(source.data[cursor:], cursor)
    parsed_data = b"".join(chunks)
    tree = parser.parse(parsed_data)
    if tree.root_node.has_error:
        return 0
    source.parsed_data = parsed_data
    source.parsed_segments = segments
    source.parsed_segment_starts = [segment[0] for segment in segments]
    source.macro_expansions = expansions
    source.tree = tree
    return len(edits)


def original_offset(source, offset):
    segments = getattr(source, "parsed_segments", ())
    if not segments:
        return offset
    index = bisect_right(source.parsed_segment_starts, offset) - 1
    if index < 0 or offset >= segments[index][1]:
        return offset
    segment = segments[index]
    start, _, origin = segment[:3]
    return origin + offset - start if len(segment) == 3 or segment[3] else origin


def expand_ident_statement_macros(sources, parser, crate_bindings, max_file_bytes):
    """Expand explicit single-ident imports without guessing macro API names.

    The body may only shadow its identifier argument. Other identifiers must be
    fully qualified paths or member names; fresh locals/repetitions are rejected.
    `$crate` receives a collision-free namespace bound to its definition source.
    """
    from syntax import child, rust_use_paths, walk
    by_path = {source.path: source for source in sources if source.language == "rust"}
    templates = {path: copy(source) for path, source in by_path.items()}
    definitions, exported = {}, {}
    for source in by_path.values():
        local, attributes = {}, []
        for node in source.tree.root_node.named_children:
            if node.type in {"line_comment", "block_comment"}:
                continue
            if node.type == "attribute_item":
                attributes.append(source.text(node))
                continue
            if node.type == "macro_definition":
                name = source.text(child(node, "name"))
                local.setdefault(name, []).append(node)
                if any(re.fullmatch(r"#\[\s*macro_export\s*\]", text) for text in attributes):
                    exported.setdefault((source.path, name), []).append(node)
            attributes = []
        definitions[source.path] = local
    notices = []
    for source in by_path.values():
        imports = {}
        for node in source.tree.root_node.named_children:
            if node.type == "use_declaration":
                for alias, path in rust_use_paths(source, child(node, "argument")):
                    imports.setdefault(alias, set()).add(path)
        nested_names = {source.text(child(node, "name")) for node in walk(source.tree.root_node)
                        if node.type == "macro_definition" and node.parent != source.tree.root_node}
        edits, generated_imports, metadata = [], {}, {}
        for invocation in walk(source.tree.root_node):
            statement = invocation.parent
            if (invocation.type != "macro_invocation" or statement is None
                    or statement.type != "expression_statement" or statement.parent.type != "block"
                    or expansion_identity(source, invocation)):
                continue
            # Local scopes can shadow imported macros. Reject rather than merge.
            name = source.text(child(invocation, "macro"))
            if name in nested_names:
                continue
            candidates = [(templates[source.path], node) for node in definitions[source.path].get(name, ())
                          if node.start_byte < invocation.start_byte]
            if not candidates:
                paths = imports.get(name, {name} if "::" in name else set())
                if len(paths) != 1:
                    continue
                for path in paths:
                    parts = path.split("::")
                    if len(parts) != 2 or parts[0] not in crate_bindings:
                        continue
                    target = templates.get(crate_bindings[parts[0]])
                    if target is not None:
                        candidates.extend((target, node) for node in exported.get((target.path, parts[1]), ()))
            if len(candidates) != 1:
                continue
            definition_source, definition = candidates[0]
            rules = [node for node in definition.named_children if node.type == "macro_rule"]
            if len(rules) != 1:
                continue
            pattern, body = child(rules[0], "left"), child(rules[0], "right")
            match = re.fullmatch(r"\(\s*\$(\w+)\s*:\s*ident\s*\)", definition_source.text(pattern))
            tokens = next((node for node in invocation.named_children if node.type == "token_tree"), None)
            arguments = [node for node in tokens.named_children if node.type not in {"line_comment", "block_comment"}] if tokens else []
            if not match or len(arguments) != 1 or arguments[0].type != "identifier":
                continue
            variable, argument = "$" + match[1], source.text(arguments[0])
            parts = list(walk(body))
            variables = [node for node in parts if node.type == "metavariable"]
            if any("repetition" in node.type for node in parts) or any(definition_source.text(node) not in {variable, "$crate"} for node in variables):
                continue
            alias = "__poc_macro_crate_" + str(len(edits))
            if alias in source.text(source.tree.root_node):
                continue
            if any(definition_source.text(node) == "$crate" for node in variables) and definition_source.path not in crate_bindings.values():
                continue
            data = getattr(definition_source, "parsed_data", definition_source.data)
            cursor, chunks = body.start_byte + 1, []
            for node in variables:
                chunks.extend([data[cursor:node.start_byte], (alias if definition_source.text(node) == "$crate" else argument).encode()])
                cursor = node.end_byte
            chunks.append(data[cursor:body.end_byte - 1])
            expanded = b"".join(chunks)
            if not statement_template_is_safe(expanded, argument, alias, parser):
                continue
            origin = original_offset(source, invocation.start_byte)
            edits.append((statement.start_byte, statement.end_byte, origin, expanded))
            generated_imports[alias] = definition_source.path
            metadata[origin] = definition_source.location(definition)
        if edits and apply_statement_edits(source, parser, edits, max_file_bytes):
            source.generated_imports = generated_imports
            source.statement_macro_definitions = metadata
            notices.append({"kind": "rust_declarative_statement_expansion", "path": source.path, "expansions": len(edits)})
    return notices


def statement_template_is_safe(expanded, argument, alias, parser):
    from syntax import child, walk
    data = b"fn __poc_template() {" + expanded + b"}"
    tree = parser.parse(data)
    if tree.root_node.has_error or len(tree.root_node.named_children) != 1:
        return False
    body = child(tree.root_node.named_children[0], "body")
    text = lambda node: data[node.start_byte:node.end_byte].decode()
    if body is None or not body.named_children:
        return False
    for node in body.named_children:
        if node.type in {"line_comment", "block_comment"}:
            continue
        pattern = child(node, "pattern")
        if node.type != "let_declaration" or pattern is None or pattern.type != "identifier" or text(pattern) != argument:
            return False
    for node in walk(body):
        if node.type in {"macro_invocation", "macro_definition", "closure_expression", "function_item", "mod_item", "use_declaration"}:
            return False
        if node.type not in {"identifier", "type_identifier"} or text(node) == argument:
            continue
        ancestor = node.parent
        while ancestor is not None and ancestor.type in {"scoped_identifier", "scoped_type_identifier", "generic_type", "generic_function"}:
            if text(ancestor).startswith(("::core::", "::std::", "::alloc::", alias + "::")):
                break
            ancestor = ancestor.parent
        else:
            return False
    return True


def apply_statement_edits(source, parser, edits, max_file_bytes):
    """Compose parser offsets while keeping original bytes, hashes and locations."""
    data = getattr(source, "parsed_data", source.data)
    if len(edits) > 256 or len(data) + sum(len(text) - (end - start) for start, end, _, text in edits) > max_file_bytes:
        return False
    old_segments = getattr(source, "parsed_segments", [(0, len(data), 0)])
    chunks, segments, expansions = [], [], []
    length, cursor = 0, 0
    def append(text, origin, linear=True):
        nonlocal length
        if text:
            segments.append((length, length + len(text), origin, linear))
            chunks.append(text)
            length += len(text)
    def unchanged(start, end):
        for segment in old_segments:
            left, right = max(start, segment[0]), min(end, segment[1])
            if left < right:
                linear = len(segment) == 3 or segment[3]
                append(data[left:right], segment[2] + left - segment[0] if linear else segment[2], linear)
    for start, end, origin, expanded in edits:
        unchanged(cursor, start)
        begin = length
        append(expanded, origin, False)
        expansions.append((begin, length, origin))
        cursor = end
    unchanged(cursor, len(data))
    parsed_data = b"".join(chunks)
    tree = parser.parse(parsed_data)
    if tree.root_node.has_error:
        return False
    def shifted(offset):
        return offset + sum(len(text) - (end - start) for start, end, _, text in edits if end <= offset)
    expansions.extend((shifted(start), shifted(end), origin) for start, end, origin in getattr(source, "macro_expansions", ()))
    source.parsed_data, source.parsed_segments = parsed_data, segments
    source.parsed_segment_starts = [segment[0] for segment in segments]
    source.macro_expansions, source.tree = sorted(expansions), tree
    return True


def statement_expansion_definition(source, node):
    for start, end, origin in getattr(source, "macro_expansions", ()):
        if start <= node.start_byte < end:
            return getattr(source, "statement_macro_definitions", {}).get(origin)
    return None


def expansion_identity(source, node):
    for start, end, origin in getattr(source, "macro_expansions", ()):
        if start <= node.start_byte < end:
            if origin in getattr(source, "statement_macro_definitions", {}):
                return f":macro@{origin}:node@{node.start_byte - start}"
            return f":macro@{origin}"
    return ""


def expansion_location(source, node):
    from model import Location
    for start, end, origin in getattr(source, "macro_expansions", ()):
        if start <= node.start_byte < end:
            row = bisect_right(source.line_starts, origin) - 1
            return Location(source.path, row + 1, origin - source.line_starts[row] + 1)
    return None

File: apps/codemap-search/src/events/source_routes/rust_macros.rs
//! Source-only expansion of the deliberately narrow declarative macro subset.
//! Macro definitions are parsed as data. No target build or procedural macro is
//! run, and every expanded span maps back to original selected source bytes.
use super::model::Location;
use super::rust_names::use_paths;
use super::syntax::*;
use std::collections::BTreeMap;
use std::sync::Arc;

#[derive(Clone, Debug)]
pub(crate) struct Segment {
    pub start: usize,
    pub end: usize,
    pub origin: usize,
    pub is_linear: bool,
}
#[derive(Clone, Debug)]
pub(crate) struct Expansion {
    pub start: usize,
    pub end: usize,
    pub origin: usize,
    pub definition: Option<Location>,
}
struct Edit {
    start: usize,
    end: usize,
    origin: usize,
    chunks: Vec<(String, usize, bool)>,
    definition: Option<Location>,
}
impl Source {
    pub fn original_offset(&self, offset: usize) -> usize {
        self.segments
            .iter()
            .rev()
            .find(|s| s.start <= offset && offset < s.end)
            .map(|s| s.origin + if s.is_linear { offset - s.start } else { 0 })
            .unwrap_or(offset)
    }
    pub fn offset_location(&self, offset: usize) -> Location {
        let offset = offset.min(self.data.len());
        let row = self
            .line_starts
            .partition_point(|start| *start <= offset)
            .saturating_sub(1);
        Location {
            path: self.path.clone(),
            line: row + 1,
            column: offset - self.line_starts[row] + 1,
        }
    }
    pub fn expansion(&self, id: NodeId) -> Option<&Expansion> {
        let start = self.nodes[id].start;
        self.expansions
            .iter()
            .find(|e| e.start <= start && start < e.end)
    }
    pub fn expansion_identity(&self, id: NodeId) -> String {
        self.expansion(id)
            .map(|e| {
                if e.definition.is_some() {
                    format!(
                        ":macro@{}:node@{}",
                        e.origin,
                        self.nodes[id].start - e.start
                    )
                } else {
                    format!(":macro@{}", e.origin)
                }
            })
            .unwrap_or_default()
    }
    fn apply_macro_edits(&mut self, mut edits: Vec<Edit>) -> bool {
        if edits.is_empty() || edits.len() > 256 {
            return false;
        }
        edits.sort_by_key(|e| e.start);
        if edits.windows(2).any(|pair| pair[0].end > pair[1].start) {
            return false;
        }
        let bytes = self.parsed_data.len() as isize
            + edits
                .iter()
                .map(|e| {
                    e.chunks.iter().map(|c| c.0.len()).sum::<usize>() as isize
                        - (e.end - e.start) as isize
                })
                .sum::<isize>();
        if bytes < 0 || bytes as usize > super::super::SOURCE_BYTES_PER_FILE {
            return false;
        }
        let old = if self.segments.is_empty() {
            vec![Segment {
                start: 0,
                end: self.parsed_data.len(),
                origin: 0,
                is_linear: true,
            }]
        } else {
            self.segments.clone()
        };
        let mut data = String::new();
        let mut segments = Vec::new();
        let mut expansions = Vec::new();
        let mut cursor = 0;
        fn append(
            data: &mut String,
            segments: &mut Vec<Segment>,
            text: &str,
            origin: usize,
            is_linear: bool,
        ) {
            if text.is_empty() {
                return;
            }
            let start = data.len();
            data.push_str(text);
            segments.push(Segment {
                start,
                end: data.len(),
                origin,
                is_linear,
            });
        }
        fn unchanged(
            source: &Source,
            old: &[Segment],
            data: &mut String,
            segments: &mut Vec<Segment>,
            start: usize,
            end: usize,
        ) {
            for segment in old {
                let left = start.max(segment.start);
                let right = end.min(segment.end);
                if left < right {
                    append(
                        data,
                        segments,
                        &source.parsed_data[left..right],
                        segment.origin
                            + if segment.is_linear {
                                left - segment.start
                            } else {
                                0
                            },
                        segment.is_linear,
                    );
                }
            }
        }
        for edit in &edits {
            unchanged(self, &old, &mut data, &mut segments, cursor, edit.start);
            let start = data.len();
            for (text, origin, linear) in &edit.chunks {
                append(&mut data, &mut segments, text, *origin, *linear);
            }
            expansions.push(Expansion {
                start,
                end: data.len(),
                origin: edit.origin,
                definition: edit.definition.clone(),
            });
            cursor = edit.end;
        }
        unchanged(
            self,
            &old,
            &mut data,
            &mut segments,
            cursor,
            self.parsed_data.len(),
        );
        let shifted = |offset: usize| -> usize {
            (offset as isize
                + edits
                    .iter()
                    .filter(|e| e.end <= offset)
                    .map(|e| {
                        e.chunks.iter().map(|c| c.0.len()).sum::<usize>() as isize
                            - (e.end - e.start) as isize
                    })
                    .sum::<isize>()) as usize
        };
        for old in &self.expansions {
            expansions.push(Expansion {
                start: shifted(old.start),
                end: shifted(old.end),
                origin: old.origin,
                definition: old.definition.clone(),
            });
        }
        expansions.sort_by_key(|e| e.start);
        let mut parser = tree_sitter::Parser::new();
        parser
            .set_language(&tree_sitter_rust::LANGUAGE.into())
            .unwrap();
        let Some(tree) = parser.parse(&data, None) else {
            return false;
        };
        if tree.root_node().has_error() {
            return false;
        }
        self.parsed_data = Arc::new(data);
        self.segments = segments;
        self.expansions = expansions;
        self.nodes.clear();
        self.collect_node(tree.root_node(), None, 0);
        self.module_bindings = self.lexical_bindings(0);
        true
    }
}
fn definition(source: &Source, id: NodeId) -> Option<(String, NodeId, Vec<NodeId>)> {
    let rules: Vec<_> = source.nodes[id]
        .children
        .iter()
        .copied()
        .filter(|n| source.nodes[*n].kind == "macro_rule")
        .collect();
    if rules.len() != 1 {
        return None;
    }
    let pattern = source.child(rules[0], &["left"])?;
    let body = source.child(rules[0], &["right"])?;
    let pattern = source.text(Some(pattern));
    let regex = regex::Regex::new(r"^\(\s*\$(\w+)\s*:\s*ident\s*\)$").unwrap();
    let captures = regex.captures(pattern)?;
    let variable = format!("${}", &captures[1]);
    let parts = source.walk(body, false);
    if parts
        .iter()
        .any(|n| source.nodes[*n].kind.contains("repetition"))
    {
        return None;
    }
    let variables: Vec<_> = parts
        .into_iter()
        .filter(|n| source.nodes[*n].kind == "metavariable")
        .collect();
    Some((variable, body, variables))
}
fn invocation_argument(source: &Source, id: NodeId) -> Option<NodeId> {
    let tokens = source.first(id, &["token_tree"])?;
    let values: Vec<_> = source.nodes[tokens]
        .children
        .iter()
        .copied()
        .filter(|n| {
            !matches!(
                source.nodes[*n].kind.as_str(),
                "line_comment" | "block_comment"
            )
        })
        .collect();
    (values.len() == 1
        && matches!(
            source.nodes[values[0]].kind.as_str(),
            "identifier" | "type_identifier"
        ))
    .then(|| values[0])
}
fn expand_items(source: &mut Source) {
    let mut definitions: BTreeMap<String, Vec<NodeId>> = BTreeMap::new();
    for &id in &source.nodes[0].children {
        if source.nodes[id].kind == "macro_definition" {
            definitions
                .entry(source.text(source.child(id, &["name"])).into())
                .or_default()
                .push(id);
        }
    }
    let mut edits = Vec::new();
    for &statement in &source.nodes[0].children {
        let invocation = if source.nodes[statement].kind == "expression_statement"
            && source.nodes[statement].children.len() == 1
        {
            source.nodes[statement].children[0]
        } else {
            statement
        };
        if source.nodes[invocation].kind != "macro_invocation" {
            continue;
        }
        let name = source.text(source.child(invocation, &["macro"]));
        let candidates: Vec<_> = definitions
            .get(name)
            .into_iter()
            .flatten()
            .copied()
            .filter(|n| source.nodes[*n].start < source.nodes[invocation].start)
            .collect();
        if candidates.len() != 1 {
            continue;
        }
        let Some((variable, body, variables)) = definition(source, candidates[0]) else {
            continue;
        };
        let Some(argument) = invocation_argument(source, invocation) else {
            continue;
        };
        if variables.is_empty() || variables.iter().any(|n| source.text(Some(*n)) != variable) {
            continue;
        }
        let mut cursor = source.nodes[body].start + 1;
        let mut chunks = Vec::new();
        for id in variables {
            chunks.push((
                source.parsed_data[cursor..source.nodes[id].start].to_owned(),
                source.original_offset(cursor),
                true,
            ));
            chunks.push((
                source.text(Some(argument)).into(),
                source.original_offset(source.nodes[argument].start),
                true,
            ));
            cursor = source.nodes[id].end;
        }
        chunks.push((
            source.parsed_data[cursor..source.nodes[body].end - 1].into(),
            source.original_offset(cursor),
            true,
        ));
        let expanded = chunks.iter().map(|c| c.0.as_str()).collect::<String>();
        let mut parser = tree_sitter::Parser::new();
        parser
            .set_language(&tree_sitter_rust::LANGUAGE.into())
            .unwrap();
        let Some(tree) = parser.parse(&expanded, None) else {
            continue;
        };
        let mut walk = tree.root_node().walk();
        if tree.root_node().has_error()
            || tree.root_node().named_child_count() == 0
            || tree.root_node().named_children(&mut walk).any(|n| {
                !matches!(
                    n.kind(),
                    "impl_item" | "line_comment" | "block_comment" | "attribute_item"
                )
            })
        {
            continue;
        }
        edits.push(Edit {
            start: source.nodes[statement].start,
            end: source.nodes[statement].end,
            origin: source.original_offset(source.nodes[invocation].start),
            chunks,
            definition: None,
        });
    }
    source.apply_macro_edits(edits);
}
fn statement_safe(text: &str, argument: &str, alias: &str) -> bool {
    let data = format!("fn __source_template() {{{text}}}");
    let Some(source) = Source::parse("__source_template.rs", &data) else {
        return false;
    };
    if source.has_parse_error {
        return false;
    }
    if source.nodes[0].children.len() != 1 {
        return false;
    }
    let Some(body) = source.child(source.nodes[0].children[0], &["body"]) else {
        return false;
    };
    if source.nodes[body].children.is_empty() {
        return false;
    }
    for &node in &source.nodes[body].children {
        if matches!(
            source.nodes[node].kind.as_str(),
            "line_comment" | "block_comment"
        ) {
            continue;
        }
        let pattern = source.child(node, &["pattern"]);
        if source.nodes[node].kind != "let_declaration"
            || pattern.is_none_or(|n| {
                source.nodes[n].kind != "identifier" || source.text(Some(n)) != argument
            })
        {
            return false;
        }
    }
    for id in source.walk(body, false) {
        let kind = source.nodes[id].kind.as_str();
        if matches!(
            kind,
            "macro_invocation"
                | "macro_definition"
                | "closure_expression"
                | "function_item"
                | "mod_item"
                | "use_declaration"
        ) {
            return false;
        }
        if !matches!(kind, "identifier" | "type_identifier") || source.text(Some(id)) == argument {
            continue;
        }
        let mut ancestor = source.nodes[id].parent;
        let mut allowed = false;
        while let Some(parent) = ancestor {
            if !matches!(
                source.nodes[parent].kind.as_str(),
                "scoped_identifier"
                    | "scoped_type_identifier"
                    | "generic_type"
                    | "generic_function"
            ) {
                break;
            }
            let text = source.text(Some(parent));
            if text.starts_with("::core::")
                || text.starts_with("::std::")
                || text.starts_with("::alloc::")
                || text.starts_with(&format!("{alias}::"))
            {
                allowed = true;
                break;
            }
            ancestor = source.nodes[parent].parent;
        }
        if !allowed {
            return false;
        }
    }
    true
}
pub(crate) fn expand_sources(
    sources: Vec<Arc<Source>>,
    bindings: &BTreeMap<String, String>,
) -> Vec<Arc<Source>> {
    let mut sources: Vec<Source> = sources.into_iter().map(|s| (*s).clone()).collect();
    for source in sources.iter_mut().filter(|s| s.language == "rust") {
        expand_items(source);
    }
    let templates = sources.clone();
    let mut definitions: BTreeMap<(usize, String), Vec<NodeId>> = BTreeMap::new();
    let mut exported: BTreeMap<(usize, String), Vec<NodeId>> = BTreeMap::new();
    for (index, source) in templates
        .iter()
        .enumerate()
        .filter(|(_, s)| s.language == "rust")
    {
        let mut is_exported = false;
        for &id in &source.nodes[0].children {
            if matches!(
                source.nodes[id].kind.as_str(),
                "line_comment" | "block_comment"
            ) {
                continue;
            }
            if source.nodes[id].kind == "attribute_item" {
                is_exported = source.text(Some(id)).contains("macro_export");
                continue;
            }
            if source.nodes[id].kind == "macro_definition" {
                let name = source.text(source.child(id, &["name"])).to_owned();
                definitions
                    .entry((index, name.clone()))
                    .or_default()
                    .push(id);
                if is_exported {
                    exported.entry((index, name)).or_default().push(id);
                }
            }
            is_exported = false;
        }
    }
    for (index, source) in sources
        .iter_mut()
        .enumerate()
        .filter(|(_, s)| s.language == "rust")
    {
        let mut imports: BTreeMap<String, Vec<String>> = BTreeMap::new();
        for &id in &source.nodes[0].children {
            if source.nodes[id].kind == "use_declaration" {
                for (alias, path) in use_paths(source, source.child(id, &["argument"]), "") {
                    imports.entry(alias).or_default().push(path);
                }
            }
        }
        let nested: std::collections::BTreeSet<_> = source
            .walk(0, false)
            .iter()
            .filter(|id| {
                source.nodes[**id].kind == "macro_definition"
                    && source.nodes[**id].parent != Some(0)
            })
            .map(|id| source.text(source.child(*id, &["name"])).to_owned())
            .collect();
        let mut edits = Vec::new();
        let mut generated = BTreeMap::new();
        for invocation in source.walk(0, false) {
            if source.nodes[invocation].kind != "macro_invocation"
                || source.expansion(invocation).is_some()
            {
                continue;
            }
            let Some(statement) = source.nodes[invocation]
                .parent
                .filter(|n| source.nodes[*n].kind == "expression_statement")
            else {
                continue;
            };
            if source.nodes[statement]
                .parent
                .is_none_or(|n| source.nodes[n].kind != "block")
            {
                continue;
            }
            let name = source.text(source.child(invocation, &["macro"]));
            if nested.contains(name) {
                continue;
            }
            let mut candidates: Vec<_> = definitions
                .get(&(index, name.into()))
                .into_iter()
                .flatten()
                .copied()
                .filter(|n| templates[index].nodes[*n].start < source.nodes[invocation].start)
                .map(|n| (index, n))
                .collect();
            if candidates.is_empty() {
                let paths = imports.get(name).cloned().unwrap_or_else(|| {
                    if name.contains("::") {
                        vec![name.into()]
                    } else {
                        Vec::new()
                    }
                });
                if paths.len() != 1 {
                    continue;
                }
                let parts: Vec<_> = paths[0].split("::").collect();
                if parts.len() != 2 {
                    continue;
                }
                let Some(path) = bindings.get(parts[0]) else {
                    continue;
                };
                let Some(target) = templates.iter().position(|s| &s.path == path) else {
                    continue;
                };
                candidates.extend(
                    exported
                        .get(&(target, parts[1].into()))
                        .into_iter()
                        .flatten()
                        .map(|n| (target, *n)),
                );
            }
            if candidates.len() != 1 {
                continue;
            }
            let (target, id) = candidates[0];
            let template = &templates[target];
            let Some((variable, body, variables)) = definition(template, id) else {
                continue;
            };
            let Some(argument) = invocation_argument(source, invocation)
                .filter(|n| source.nodes[*n].kind == "identifier")
            else {
                continue;
            };
            let argument = source.text(Some(argument));
            if variables.iter().any(|n| {
                !matches!(template.text(Some(*n)), "$crate") && template.text(Some(*n)) != variable
            }) {
                continue;
            }
            let alias = format!("__source_macro_crate_{}", edits.len());
            if source.parsed_data.contains(&alias) {
                continue;
            }
            if variables
                .iter()
                .any(|n| template.text(Some(*n)) == "$crate")
                && !bindings.values().any(|p| p == &template.path)
            {
                continue;
            }
            let mut expanded = String::new();
            let mut cursor = template.nodes[body].start + 1;
            for variable in variables {
                expanded.push_str(&template.parsed_data[cursor..template.nodes[variable].start]);
                expanded.push_str(if template.text(Some(variable)) == "$crate" {
                    &alias
                } else {
                    argument
                });
                cursor = template.nodes[variable].end;
            }
            expanded.push_str(&template.parsed_data[cursor..template.nodes[body].end - 1]);
            if !statement_safe(&expanded, argument, &alias) {
                continue;
            }
            let origin = source.original_offset(source.nodes[invocation].start);
            let definition = Some(template.location(id));
            edits.push(Edit {
                start: source.nodes[statement].start,
                end: source.nodes[statement].end,
                origin,
                chunks: vec![(expanded, origin, false)],
                definition,
            });
            generated.insert(alias, template.path.clone());
        }
        if source.apply_macro_edits(edits) {
            source.generated_imports = generated;
        }
    }
    sources.into_iter().map(Arc::new).collect()
}
